### PR TITLE
kvm: Handle storage issue on NFS/KVM in multiple ways

### PR DESCRIPTION
**Pull request by ustcweizhou**
### Description

This PR provides multiple options to handle storage issue on kvm.

FS: to be added

Two main parts
(1) kvm: Handle storage issue on NFS/KVM in multiple ways : hardreset (default), noaction, destroyvms, stopagent
(2) kvm: kvm: Schedule investigate tasks for disconnected hosts

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [X] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

### CODE REVIEW

**Comment by weizhouapache** [comment]
related to #2722 #2890 #2984 #4586

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by GabrielBrascher** [comment]
@ustcweizhou thanks for the PR. From what I understood (really quick glance of eyes) this avoid also situations of HA monitor a removed NFS storage. Am I right?

I have seen some weird issues where a Cluster that had NFS and removes it (standard storage removal process `Up > Maintenance > Removed`) but still HA expects NFS pool to be "Healthy" and simply Fence all hosts on the cluster.

**Comment by nvazquez** [comment]
Hi @ustcweizhou can you please resolve the conflicts?

File: plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/KVMHAMonitor.java [hunk 1]
```diff
@@ -77,84 +92,148 @@ public NfsStoragePool getStoragePool(String uuid) {
...
 
-    private class Monitor extends ManagedContextRunnable {
-
-        @Override
-        protected void runInContext() {
-            synchronized (_storagePool) {
-                Set<String> removedPools = new HashSet<String>();
-                for (String uuid : _storagePool.keySet()) {
-                    NfsStoragePool primaryStoragePool = _storagePool.get(uuid);
+    protected class CheckPoolThread extends Thread {
```
Comment (DaanHoogland) [inline plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/KVMHAMonitor.java:95]: why not make this a ManagemedContextRunnable as well?

**Review by DaanHoogland — COMMENTED** [review]
I wonder about the use of Thread directly, @weizhouapache. Any particular reason?

**Comment by rhtyd** [comment]
Ping @weizhouapache can you fix the conflicts, is this essential for 4.16?

File: engine/orchestration/src/main/java/com/cloud/agent/manager/AgentManagerImpl.java [hunk 2]
```diff
@@ -238,6 +261,9 @@ public boolean configure(final String name, final Map<String, Object> params) th
 
         _monitorExecutor = new ScheduledThreadPoolExecutor(1, new NamedThreadFactory("AgentMonitor"));
 
+        _scanHostsExecutor =  new ScheduledThreadPoolExecutor(1, new NamedThreadFactory("HostsScanner"));
+        _investigatorExecutor = new ScheduledThreadPoolExecutor(InvestigateDisconnectedHostsPoolSize.value(), new NamedThreadFactory("DisconnectHostsInvestigator"));
```
Comment (sureshanaparti) [inline engine/orchestration/src/main/java/com/cloud/agent/manager/AgentManagerImpl.java:265]:
```suggestion
        _investigatorExecutor = new ScheduledThreadPoolExecutor(InvestigateDisconnectedHostsPoolSize.value(), new NamedThreadFactory("DisconnectedHostsInvestigator"));
```

File: engine/orchestration/src/main/java/com/cloud/agent/manager/AgentManagerImpl.java [hunk 3]
```diff
@@ -773,6 +803,8 @@ public boolean stop() {
 
         _connectExecutor.shutdownNow();
         _monitorExecutor.shutdownNow();
+        _investigatorExecutor.shutdownNow();
+
```
Comment (sureshanaparti) [inline engine/orchestration/src/main/java/com/cloud/agent/manager/AgentManagerImpl.java:807]: `_scanHostsExecutor.shutdownNow()` missing

File: engine/orchestration/src/main/java/com/cloud/agent/manager/AgentManagerImpl.java [hunk 4]
```diff
@@ -203,6 +218,14 @@
     protected final ConfigKey<Boolean> CheckTxnBeforeSending = new ConfigKey<Boolean>("Developer", Boolean.class, "check.txn.before.sending.agent.commands", "false",
             "This parameter allows developers to enable a check to see if a transaction wraps commands that are sent to the resource.  This is not to be enabled on production systems.", true);
 
+    protected final ConfigKey<Boolean> InvestigateDisconnectedHosts = new ConfigKey<>("Advanced", Boolean.class, "investigate.disconnected.hosts",
+        "false", "Determines whether to investigate VMs on disconnected hosts", false);
+    protected final ConfigKey<Integer> InvestigateDisconnectedHostsInterval = new ConfigKey<>("Advanced", Integer.class, "investigate.disconnected.hosts.interval",
+        "300", "The time (in seconds) between VM investigation on disconnected hosts.", false);
+    protected final ConfigKey<Integer> InvestigateDisconnectedHostsPoolSize = new ConfigKey<Integer>("Advanced", Integer.class, "investigate.disconnected.hosts.pool.size", "10",
+            "Default pool size to investigate disconnected hosts", false);
```
Comment (sureshanaparti) [inline engine/orchestration/src/main/java/com/cloud/agent/manager/AgentManagerImpl.java:226]:
```suggestion
            "The thread pool size to investigate disconnected hosts", false);
```

**Comment by sureshanaparti** [comment]
@weizhouapache can you address outstanding remarks, and fix the conflicts.

**Comment by weizhouapache** [comment]
@rhtyd @sureshanaparti 
this PR has many conflicts with #5239 which has been merged into master.
I think it is better to re-target to 4.17

what do you think ? @ravening @soreana

File: plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/KVMHAMonitor.java (hunk 1 above)
Comment (weizhouapache) [reply plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/KVMHAMonitor.java:95]: ManagedContextRunnable and Thread are both implemetation of Runnable interface.
Thread works in testing. 

ManagedContextRunnable should also be ok. The following might work as well
```
protected class CheckPoolThread implements Runnable {
```

**Comment by soreana** [comment]
@weizhouapache I'm agree 👍

**Comment by soreana** [comment]
@weizhouapache I hope you are doing well :)
I'm going to work on this pr. How can I reproduce it?

**Comment by weizhouapache** [comment]
> @weizhouapache I hope you are doing well :) I'm going to work on this pr. How can I reproduce it?

@soreana 
what do you want to reproduce ? this is an enhancement, not a bug.

**Comment by soreana** [comment]
> > @weizhouapache I hope you are doing well :) I'm going to work on this pr. How can I reproduce it?
> 
> @soreana what do you want to reproduce ? this is an enhancement, not a bug.

Some test cases to see if it works or not would be enough.

**Comment by weizhouapache** [comment]
> > > @weizhouapache I hope you are doing well :) I'm going to work on this pr. How can I reproduce it?
> > 
> > 
> > @soreana what do you want to reproduce ? this is an enhancement, not a bug.
> 
> Some test cases to see if it works or not would be enough.

@soreana 
I think you can get more info on your wiki/jira, etc.

**Comment by ravening** [comment]
@weizhouapache I created a new pr for this in https://github.com/apache/cloudstack/pull/5783

**Comment by sureshanaparti** [comment]
> @weizhouapache I created a new pr for this in #5783

@weizhouapache @ravening is this PR no longer relevant (as the changes are ported to #5783)?

**Comment by ravening** [comment]
> > @weizhouapache I created a new pr for this in #5783
> 
> @weizhouapache @ravening is this PR no longer relevant (as the changes are ported to #5783)?

@sureshanaparti yes. this has conflicts.. i have resolved the conflicts and created new pr

**Pull request closed without merging**